src/synology_dsm/api/core/external_usb.py

```diff
@@ -0,0 +1,245 @@
+        size_total = self.partitions_all_size_total(human_readable=False)
+        size_used = self.partitions_all_size_used(human_readable=False)
+
+        if size_used >= 0 and size_total and size_total > 0:
+            return round((float(size_used) / float(size_total)) * 100.0, 1)
+        return None
+
+
+class SynoUSBStoragePartition:
+    """A representation of a parition of an external USB storage device."""
+
+    def __init__(self, data):
+        """Initialize a partition object of an external USB storage device."""
+        self._data = data
+
+    @property
+    def fstype(self):
+        """Return the dev_fstype for the partition."""
+        return self._data["dev_fstype"]
+
+    @property
+    def filesystem(self):
+        """Return the filesystem for the partition."""
+        return self._data["filesystem"]
+
+    @property
+    def name_id(self):
+        """Return the name_id for the partition."""
+        return self._data["name_id"]
+
+    @property
+    def partition_title(self):
+        """Return the title for the partition."""
+        return self._data["partition_title"]
+
+    @property
+    def share_name(self):
+        """Return the share name for the partition."""
+        return self._data["share_name"]
+
+    @property
+    def status(self):
+        """Return the status for the partition."""
+        return self._data["status"]
+
+    def partition_size_total(self, human_readable=False):
+        """Total size of the partition."""
+        # API returns property as empty string if a partition has no size
+        size_total = self._data["total_size_mb"]
+        if size_total == "":
+            return None
+        size_total = SynoFormatHelper.megabytes_to_bytes(int(size_total))
+        if human_readable:
+            return SynoFormatHelper.bytes_to_readable(size_total)
+        return size_total
+
+    def partition_size_used(self, human_readable=False):
+        """Used size of the partition."""
+        # API does not return property if a partition has no size
+        if "used_size_mb" not in self._data:
+            return None
+        size_used = self._data["used_size_mb"]
+        if size_used == "":
+            return None
+        size_used = SynoFormatHelper.megabytes_to_bytes(int(size_used))
+        if human_readable:
+            return SynoFormatHelper.bytes_to_readable(size_used)
+        return size_used
+
+    @property
+    def partition_percentage_used(self):
+        """Used size in percentage of the partition."""
+        size_total = self.partition_size_total()
+        size_used = self.partition_size_used()
+        if size_total is None or size_used is None:
+            return None
+
+        if size_used >= 0 and size_total and size_total > 0:
+            return round((float(size_used) / float(size_total)) * 100.0, 1)
+        return None
+
+    def is_formatted(self):
+        """Is the partition formatted."""
+        return self._data["total_size_mb"] != ""
```

Comment:
i guess, partitions with unknown filesystem will also report no total size, so `is_formatted` may be misleading

Reply:
Hi @mib1185 
I created a partition using Apple's APFS on a stick and even though the fstype was listed as unknown, it still reported the size.  I'm just guessing, but I think the partition size is stored in the partition table data - MBR or GPT/GUID.  So even if it can't read the partition it can still know the size.  I'll keep trying to break it.  

Here's what the API returned:
```
{
  "data": {
    "devices": [
      ...
      {
        "dev_id": "usb5",
        "dev_title": "USB Disk 5",
        "dev_type": "usbDisk",
        "formatable": true,
        "partitions": [
          {
            "dev_fstype": "vfat",
            "filesystem": "FAT32",
            "name_id": "usb5p1",
            "partition_title": "USB Disk 5 Partition 1",
            "share_name": "usbshare5-1",
            "status": "normal",
            "total_size_mb": 197,
            "used_size_mb": 0
          },
          {
            "dev_fstype": "hfsplus",
            "filesystem": "hfsplus",
            "name_id": "usb5p2",
            "partition_title": "USB Disk 5 Partition 2",
            "share_name": "usbshare5-2",
            "status": "normal",
            "total_size_mb": 1796,
            "used_size_mb": 24
          },
          {
            "dev_fstype": "unknown",
            "filesystem": "",
            "name_id": "usb5p3",
            "partition_title": "USB Disk 5 Partition 3",
            "share_name": "",
            "status": "normal",
            "total_size_mb": 1795,
            "used_size_mb": 0
          }
        ],
        "producer": "SanDisk Corp.",
        "product": "Cruzer Titanium+",
        "progress": "",
        "status": "normal",
        "total_size_mb": 3919
      },
      ...
    ]
  },
  "success": true
}
```

Or maybe instead of is_formatted there could be something that indicates that it has a recognized filesystem?  It's not necessary, more of a convenience to know if the partition is initialized or likely usable by the NAS?

Reply:
the question is, in which situations is `total_size_mb` returned as an empty string (`""`) :thinking: 
if it is just about to check if the partition is accessable, we can check if `filesystem` and/or a `share_name` (_mounted folder on NAS_) is provided? What do you think?

Reply:
After a little more experimenting, I like going by the share_name and filesystem if we want to have something indicating the partition is accessible.  When erasing the drive on the Mac, there was a third option besides MBR and GPT called Apple Partition Map.  When I used that, it ended up like in the screenshot/data below.  It appears it needs to have both a filesystem and share_name to be usable, but what to call it?  
- is_available
- is_mounted
- is_shared
- is_supported


Here's a screenshot of what is displayed in DSM along with the data returned by the API.  Maybe instead of a single property, have separate properties for is_shared and is_filesystem_supported to match how it's displayed in DSM.  But you could already determine that by just checking filesystem or share_name.  Perhaps it's like device_ids and isn't necessary at all.
![image](https://user-images.githubusercontent.com/5054745/219974137-f8f1ccfd-3e87-4a9f-9b00-09b14d8103a6.png)

```
      {
        "dev_id": "usb5",
        "dev_title": "USB Disk 5",
        "dev_type": "usbDisk",
        "formatable": true,
        "partitions": [
          {
            "dev_fstype": "unknown",
            "filesystem": "",
            "name_id": "usb5p1",
            "partition_title": "USB Disk 5 Partition 1",
            "share_name": "",
            "status": "normal",
            "total_size_mb": 0,
            "used_size_mb": 0
          },
          {
            "dev_fstype": "unknown",
            "filesystem": "",
            "name_id": "usb5p2",
            "partition_title": "USB Disk 5 Partition 2",
            "share_name": "",
            "status": "normal",
            "total_size_mb": 128,
            "used_size_mb": 0
          },
          {
            "dev_fstype": "hfsplus",
            "filesystem": "hfsplus",
            "name_id": "usb5p3",
            "partition_title": "USB Disk 5 Partition 3",
            "share_name": "usbshare5-3",
            "status": "normal",
            "total_size_mb": 3792,
            "used_size_mb": 25
          },
          {
            "dev_fstype": "unknown",
            "filesystem": "",
            "name_id": "usb5p4",
            "partition_title": "USB Disk 5 Partition 4",
            "share_name": "",
            "status": "normal",
            "total_size_mb": 0,
            "used_size_mb": 0
          }
        ],
        "producer": "SanDisk Corp.",
        "product": "Cruzer Titanium+",
        "progress": "",
        "status": "normal",
        "total_size_mb": 3919
      },
```

Reply:
i think `is_mounted` should reflect if a `share_name` is present and `is_supported` in case `filesystem` is not empty